### PR TITLE
Access database and other files through features, for simpler configuration

### PR DESCRIPTION
**Pull request by tornaria**
This PR reworks the way sagemath access database files, and some other external files. The end result is that the runtime detection of databases as features matches exactly what will be used. Moreover, this leads to an easy implementation of search paths, meaning we don't need configuration as long as the files are in a few standard system locations.

Notes on implementation:
- First every user of these files access them via features, and never directly through `sage.env`, since the former is dynamic and the latter is static.
- Then we add a variable `SAGE_DATA_PATH` which is a colon separated search path for databases. The default should work for sage-the-distro and most distros (think `/usr/share/sagemath:/usr/share` for a system install on `/usr`).
- Now a database, say `cremona` is searched on `$p/cremona` for each `$p` in `SAGE_DATA_PATH`.
- I also added `$DOT_SAGE/db` first in the search path, this makes it easy for a user to install a missing database (or update an old one).
- Finally, as an example I did something similar to find `JmolData.jar` and `three.min.js`. The former is searched in `$SAGE_SHARE/sagemath/jmol` and `$SAGE_SHARE/jmol` and the latter also in `$SAGE_SHARE/jupyter/nbextensions/threejs-sage` which seems to make sense.

Other files can be done later (e.g. mathjax). In principle anything that has a default value in `sage.env` or `sage_conf` should benefit from using a search path and removing the default value.

With this PR, I can run stock sagemath without any `sage_conf.py`.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have created tests covering the changes.

### :hourglass: Dependencies

### CODE REVIEW

**Comment by tornaria** [comment]
What's with this:
```
UserWarning: Feature database_cremona_mini_ellcurve is declared optional, but it is provided by elliptic_curves, which is declared standard in SAGE_ROOT/build/pkgs
```
I can't reproduce.

File: src/sage/graphs/strongly_regular_db.pyx
Comment (mkoeppe) [inline src/sage/graphs/strongly_regular_db.pyx]:
```suggestion
    cdef int n = <int> floor(sqrt(n_2_p_1 - 1))
```

accidental change?

**Comment by mkoeppe** [comment]
Very nice. I'll take a closer look later.

File: src/sage/features/databases.py
Comment (mkoeppe) [inline src/sage/features/databases.py:113]:
```suggestion
                            spkg="elliptic_curves",
                            type='standard',
```

File: src/sage/features/databases.py
Comment (mkoeppe) [inline src/sage/features/databases.py:143]:
```suggestion
                            spkg="graphs",
                            type='standard',
```

File: src/sage/features/jmol.py
Comment (mkoeppe) [inline src/sage/features/jmol.py:37]:
```suggestion
            spkg="jmol",
            type='standard',
```

File: src/sage/features/threejs.py
Comment (mkoeppe) [inline src/sage/features/threejs.py:40]:
```suggestion
            spkg="threejs",
            type='standard',
```

File: src/sage/features/__init__.py
Comment (mkoeppe) [inline src/sage/features/__init__.py]: is this used somewhere?
splitting on colon is a potential porting hazard when native Windows paths get involved

File: src/sage/graphs/strongly_regular_db.pyx
Comment (tornaria) [reply src/sage/graphs/strongly_regular_db.pyx]: I did it to silence pycodestyle E225. Is it incorrect cython? Is there a way to tell pycodestyle to respect cython extensions like this while still complaining about python style? Is there an easy way to run `pycodestyle-minimal` on individualy `*.pyx` files?
```
$ tox -e pycodestyle-minimal -- src/sage/graphs/strongly_regular_db.pyx
...
sage/graphs/strongly_regular_db.pyx:1674:22: E225 missing whitespace around operator
...
```
Running `tox -e pycodestyle-minimal` does not check E225 on pyx files, but it takes a long time.

In fact, passing a `*.pyx` filename results in pycodestyle running twice, it seems, but the first run should be excluded for `*.pyx` files.

Still, E225 is an useful warning for `*.pyx` files, except for those false positives.  Should I tag them instead? For instance, in this file there are 4 true positives on E225.

File: src/sage/features/databases.py
Comment (tornaria) [reply src/sage/features/databases.py:113]: `type=standard` is a sage-the-distro thing. Why do we need to add this information here as well? What is it useful for here?

File: src/sage/features/__init__.py
Comment (tornaria) [reply src/sage/features/__init__.py]: No, I don't think I'm using it right now, but might use it later so I prefer to state the intention now... Using `:` as separators is standard, and we need this for `SAGE_DATA_PATH`. We could use `os.pathsep`, but I think ":" is already used explicitly in a some places (e.g. `Executable.absolute_filename()`).

Internally we can use lists, but this is meant to communicate through environment vaiables so there is need to serialize.

How does WSL solve this?

File: src/sage/graphs/strongly_regular_db.pyx
Comment (mkoeppe) [reply src/sage/graphs/strongly_regular_db.pyx]: Just use the right linter for Cython. https://doc.sagemath.org/html/en/developer/tools.html#cython-lint

File: src/sage/features/__init__.py
Comment (mkoeppe) [reply src/sage/features/__init__.py]: I don't see why this would have to be handled on the level of the input args of the `Feature`. The input args have Python types.

File: src/sage/features/databases.py
Comment (mkoeppe) [reply src/sage/features/databases.py:113]: It suppresses the warning that you complained about

File: src/sage/graphs/strongly_regular_db.pyx
Comment (tornaria) [reply src/sage/graphs/strongly_regular_db.pyx]: > Just use the right linter for Cython. https://doc.sagemath.org/html/en/developer/tools.html#cython-lint

Thanks, but this is not what CI tests, is it?

File: src/sage/features/databases.py
Comment (tornaria) [reply src/sage/features/databases.py:113]: Sure, but it seems unnecessary to repeat this information here. In the end, all that matters at runtime is whether the feature `is_present()` or not. Sad consolation that a feature `is_standard()` if it's not present.

File: src/sage/features/__init__.py
Comment (tornaria) [inline src/sage/features/__init__.py]: There's one test failure here because I forgot to change this test after I removed the split at ":" feature for `search_path`.

I'll fix it after current CI finishes with this:
```suggestion
            ....:                 search_path=("/dev", "/bin", "/usr"))
```

**Comment by tornaria** [comment]
I think this is ready to go. Only failure is in codecov/patch, but for some of the new lines I have no way to test them in CI (need optional packages, etc).

**Comment by tornaria** [comment]
I moved `_required_threejs_version` to a method of the threejs feature, since `sage.repl.rich_output.display_manager` cannot be imported at install time.

**Comment by mkoeppe** [comment]
Overall, looks good to me. 

The solution for replacing the `COMBINATORIAL_DESIGN_DATA_DIR` seems fine to me. I don't have a preference between this approach and @orlitzky's approach in #37025 using a new repo. Merging the present branch does not conflict with implementing @orlitzky's approach if so decided later.

I'll hold off with setting to positive review for a moment so that other downstream packagers can comment whether this meets their needs. Feel free to set to positive review if there's no reaction.

**Review by mkoeppe — APPROVED** [review]

File: src/sage/features/threejs.py
Comment (kiwifb) [inline src/sage/features/threejs.py:35]: I love the design idea and I am all for it. Here is more of a comment that we may want to change some the packaging we have for packages like threejs to match better one of those sensible folders out of the box.
I am installing threejs-sage-extension straight from pypi, 0 patch or changes and this is what I get
```
/usr/share/sage
/usr/share/sage/threejs
/usr/share/sage/threejs/r122
/usr/share/sage/threejs/r122/three.min.js
/usr/share/sage/threejs/version
```
We are still at `sage` instead of `sagemath`. I would love for the package to install in the jupyter nbextension folder out of the box which feels right and sensible. Right now, none of these paths, that are looking sensible, will work out of the box without a little bit of work on threejs-sage-extension.

Comment (tornaria) [reply src/sage/features/threejs.py:35]: Good point, so:
 1. I'll add `SAGE_SHARE/sage/threejs` to the search path above, does that sound right?
 2. I agree that it would make more sense for the threejs sage extension to be installed into jupyter folder, i.e. as above `SAGE_SHARE/jupyter/nbextensions/threejs-sage`.
 3. In the same  I think something like this could eventually be done:
```diff
--- a/build/pkgs/threejs/spkg-install.in
+++ b/build/pkgs/threejs/spkg-install.in
@@ -1,2 +1,2 @@
-sdh_install    src/version "${SAGE_SHARE}/threejs-sage/"
-sdh_install -T src/build   "${SAGE_SHARE}/threejs-sage/$(cat src/version)"
+sdh_install    src/version "${SAGE_SHARE}/jupyter/nbextensions/threejs-sage/"
+sdh_install -T src/build   "${SAGE_SHARE}/jupyter/nbextensions/threejs-sage/$(cat src/version)"
```

I wonder if it would be possible for sagemath to use https://pypi.org/project/pythreejs/ instead of this custom threejs.

Comment (tornaria) [reply src/sage/features/threejs.py:35]: @kiwifb Where is your `threejs-sage-extension` coming from? The one I found in pypi is named `jupyter-threejs-sage`. Current version is 123 and using pip to install fails with
 
   FileNotFoundError: [Errno 2] No such file or directory: 'version'

There doesn't seem to be version 122, what worked for me is to install `jupyter-threejs-sage=122b1` which installed `.../share/jupyter/nbextensions/threejs-sage/r122/three.min.js`.

**Comment by orlitzky** [comment]
I strongly believe that separate packages with static sagelib dependencies and discoverability via python is a better design, but this doesn't make anything worse at the moment.

**Comment by tornaria** [comment]
> Can you please include my improvements from https://github.com/orlitzky/mols-handbook-data/commits/master/ since this ticket will be used as an excuse not to merge them?

Sure, should I just place your `__init__.py` file as `src/sage/combinat/designs/mols_handbook_data.py` and use it through `mols_handbook_data.lower_bound()` ?

This way we keep your documentation and testing.

**Comment by orlitzky** [comment]
> Sure, should I just place your `__init__.py` file as `src/sage/combinat/designs/mols_handbook_data.py` and use it through `mols_handbook_data.lower_bound()` ?

Yeah that's fine with me. A few doctests might need to be tweaked to use `sage: ` rather than `>>> `, and the docstring for `lower_bound()` is [NumPy-style](https://www.sphinx-doc.org/en/master/usage/extensions/example_numpy.html) rather than Sage-style. Otherwise it's just a file full of numbers.

**Comment by tornaria** [comment]
> > Sure, should I just place your `__init__.py` file as `src/sage/combinat/designs/mols_handbook_data.py` and use it through `mols_handbook_data.lower_bound()` ?
> 
> Yeah that's fine with me. A few doctests might need to be tweaked to use `sage: ` rather than `>>> `, and the docstring for `lower_bound()` is [NumPy-style](https://www.sphinx-doc.org/en/master/usage/extensions/example_numpy.html) rather than Sage-style. Otherwise it's just a file full of numbers.

Done. Feel free to make a PR against my branch if you think anything else needs changing, or just make a follow-up PR on top of this.

**Comment by tornaria** [comment]
> I strongly believe that separate packages with static sagelib dependencies and discoverability via python is a better design, but this doesn't make anything worse at the moment.

I think using separate packages is fine if there's an active upstream that cares about the package. What you are suggesting is that sagemath becomes the upstream for a package that (in principle) only sagemath will use and for which there is no technical advantage for splitting.

Being upstream for a few packages is not working well for sagemath, and it doesn't scale. You make the package to start with, but are you willing to maintain it? Do you even have the bandwidth to maintain all of this? For instance, you created conway-polynomials and got it used in sagemath, but there are two open PRs (from *before* this was merged into sagemath) that didn't even get a "not interested" comment.

File: src/sage/features/threejs.py
Comment (kiwifb) [reply src/sage/features/threejs.py:35]: Your comments surprised me so I doubled checked. I indeed use 122b1. You did install using pip? I will inspect what happens when I install.

Comment (kiwifb) [reply src/sage/features/threejs.py:35]: I made a stupid mistake. For some reason I was looking at the files of `threejs-sage-extension`. Wrong package. Sometimes, I am wondering if I am becoming senile. Now the curious bit to me is that I do not have `threejs-sage-extension` installed and everything doctest fine.

Comment (tornaria) [reply src/sage/features/threejs.py:35]: Indeed:
```
$ python -m venv threejs-sage
$ . threejs-sage/bin/activate
(threejs-sage) $ pip install jupyter-threejs-sage==122b1
Collecting jupyter-threejs-sage==122b1
  Using cached jupyter_threejs_sage-122b1-py3-none-any.whl
Installing collected packages: jupyter-threejs-sage
Successfully installed jupyter-threejs-sage-122b1
(threejs-sage) $ pip show --files jupyter-threejs-sage
Name: jupyter-threejs-sage
Version: 122b1
...
Files:
  ../../../share/jupyter/nbextensions/threejs-sage/r122/three.min.js
  jupyter_threejs_sage-122b1.dist-info/INSTALLER
  jupyter_threejs_sage-122b1.dist-info/LICENSE
  jupyter_threejs_sage-122b1.dist-info/METADATA
  jupyter_threejs_sage-122b1.dist-info/RECORD
  jupyter_threejs_sage-122b1.dist-info/REQUESTED
  jupyter_threejs_sage-122b1.dist-info/WHEEL
  jupyter_threejs_sage-122b1.dist-info/top_level.txt
  jupyter_threejs_sage/__init__.py
  jupyter_threejs_sage/__pycache__/__init__.cpython-312.pyc
  jupyter_threejs_sage/static/r122/three.min.js
(threejs-sage) $ 
```

Note that:
 - pipy only has 122b1 and 123 (the latter is broken)
 - https://github.com/sagemath/threejs-sage/ shows `r130` as the latest release (jul 7, 2021)
 - repology claims `r124` is the newest version for https://repology.org/project/threejs-sage/ and `r122` for https://repology.org/project/threejs/
 - updating is a pain since the filepaths are versioned (wtf?)

:shrug:

Comment (kiwifb) [reply src/sage/features/threejs.py:35]: I am also increasingly not a morning person. Should re-read the whole thing. Yes I have packages for `threejs-sage-extension` which installs where I said it does, and the installation is manual. But I was mistakenly looking at the ebuild of `jupyter-threejs-sage` which is coming from pypi and would install nicely as an extension.

Which makes me wonder, was one meant to replace the other? It looks like they have payload with the same name, just different location. `/usr/share/sage/threejs/r122/three.min.js` for one and `/usr/share/jupyter/nbextensions/threejs-sage/r122/three.min.js` for the other.

**Comment by orlitzky** [comment]
> I think using separate packages is fine if there's an active upstream that cares about the package. What you are suggesting is that sagemath becomes the upstream for a package that (in principle) only sagemath will use and for which there is no technical advantage for splitting.

SageMath is also upstream for it when you include it with sagelib. The technical advantage for the combinatorial designs is minor (for example, saving space in most peoples' git checkouts) but nonzero. For the other database packages, the advantages are greater: not having to define `SAGE_SHARE` or guess at paths, not having coupling in both directions between sage and its database packages.

> Being upstream for a few packages is not working well for sagemath, and it doesn't scale. You make the package to start with, but are you willing to maintain it? Do you even have the bandwidth to maintain all of this? For instance, you created conway-polynomials and got it used in sagemath, but there are two open PRs (from _before_ this was merged into sagemath) that didn't even get a "not interested" comment.

I moved the repo under the sagemath organization so that I'm not a single point of failure, but if there's a group of people who get sagemath notifications by default, I'm not in it. You haven't gotten a response not because I'm ignoring them, but because I didn't know they were there.

File: src/sage/features/threejs.py
Comment (kiwifb) [reply src/sage/features/threejs.py:35]: And those files are identical too.

**Comment by tornaria** [comment]
> SageMath is also upstream for it when you include it with sagelib.

Yes but it's just a file out of 10k.

> The technical advantage for the combinatorial designs is minor (for example, saving space in most peoples' git checkouts) but nonzero.

Not significant, moreover all the files `build/pkgs/combinatorial_designs` takes more blocks than the actual file (!). But on the other side of the coin, you add another gh repo, another upstream, another pypi project, another package for each distribution, etc.

> For the other database packages, the advantages are greater: not having to define `SAGE_SHARE` or guess at paths, not having coupling in both directions between sage and its database packages.

Sure: having a proper "python" package instead of an ad-hoc "spkg" package seems nicer, but it still has a cost, and an "ad-hoc" python package may not be enough benefit to offset the cost. At least updating a file in the main repo is something that we know how to do, but managing several packages and pypi is something we do badly.

Examples: cypari2 (I think) needed to be updated but noone could do it in pypi; threejs-sage is a mess and there is an upstream but sage is too special to use that so fork it and then have a mishmash of versions between spkg / gh / pypi / upstream, moreover having a separate package (forked, under full sagemath control) is useless since the filepaths have an exact version so the package can't really be updated independent of sagemath.

Since I'm ranting: we have sage_setup split off (why?) but we require *exactly* the same version, what is that good for? I have sage_setup 10.2 in my system and I can't use it to build 10.3.beta4, wtf, etc, etc.

> I moved the repo under the sagemath organization so that I'm not a single point of failure, but if there's a group of people who get sagemath notifications by default, I'm not in it. You haven't gotten a response not because I'm ignoring them, but because I didn't know they were there.

I'm not blaming you, and the issue is not new nor specific to this package. I have other PRs in the same status (e.g. https://github.com/sagemath/cysignals/pull/192). Another one is https://github.com/sagemath/cypari2/pull/138, and before you say this is merged, look again because here it's not: https://pypi.org/project/cypari2/.

File: src/sage/features/threejs.py
Comment (tornaria) [reply src/sage/features/threejs.py:35]: Ok, so your `threejs-sage-extension` package is just like my `threejs-sage` package and sage-the-distro `threejs` package, and arch `threejs-sage` package, which gets a tarball (presumably from github repo) and places the files by hand in a certain location.  It's just that you place it in `/usr/share/sage/threejs`, I doo in `/usr/share/sagemath/threejs-sage` arch does it in `/usr/share/threejs-sage` and sage-the-distro does it in `SAGE_SHARE/threejs-sage`. And the pypi package looks in jupyter nbextensions directory.

Now all of those are caught by the search path above, and I hope @orlitzky will forgive us for looking into 4 different places for the file. And all this trouble is very relevant to our discussion about external packages that are created without taking into account the cost of future maintenance (which noone does, so the maintenance cost carries over to whoever has to sort out the mess when it blows up).

I'll mark this as resolved.

Is this working? Can we flag this as positive review now?

**Comment by orlitzky** [comment]
> moreover all the files `build/pkgs/combinatorial_designs` takes more blocks than the actual file (!)

Let it be recorded that on this day, the Eleventh of January, in the year 2024, Gonzalo Tornaria cared about the sage distribution :)

I'm looking forward to a time when we can check out sagelib and work on it without all the `build/` stuff.

No argument that we should have a better plan for maintaining other sagemath and pypi packages.

File: src/sage/features/threejs.py
Comment (orlitzky) [reply src/sage/features/threejs.py:35]: It's not about the number of locations. The ThreeJS data is unusable without sage and sage is unusable without the ThreeJS data. Packaging either individually becomes awkward, and when you do package them, they're coupled because they have to agree on the data directory.

Contrast with a python package where the search path is implicitly determined by the --libdir used to configure python. Neither sage nor the database package needs to know that libdir for them to work together. Sage depends on conway-polynomials, but conway-polynomials is now completely independent of sage. The dependency graph has become simpler.

I've been trying to get Sage into Gentoo for 15 years and I'd really like to be able to cross these database packages off my TODO list one-at-a-time. Making them independently useful and of high quality makes that easier, practically and politically.

**Comment by tornaria** [comment]
> > moreover all the files `build/pkgs/combinatorial_designs` takes more blocks than the actual file (!)
> 
> Let it be recorded that on this day, the Eleventh of January, in the year 2024, Gonzalo Tornaria cared about the sage distribution :)

LOL

> I'm looking forward to a time when we can check out sagelib and work on it without all the `build/` stuff.

I'm about to switch my void package to use `sagemath-standard` from pypi. At this point I think it's doable.

Now *that* is a move I would support: splitting `sage-the-distro` from `sagemath-standard`. Have `sage-the-distro` use `sagemath-standard` as upstream, with no special privileges (other than maybe more testing) wrt other distros.

> No argument that we should have a better plan for maintaining other sagemath and pypi packages.

Yes, and the better way for that is to replace sagemath-upstream packages with third-party upstream packages, when possible. Keep in mind the three R (reduce-reutilize-recycle).

File: src/sage/features/threejs.py
Comment (tornaria) [reply src/sage/features/threejs.py:35]: > It's not about the number of locations. The ThreeJS data is unusable without sage and sage is unusable without the ThreeJS data. Packaging either individually becomes awkward, and when you do package them, they're coupled because they have to agree on the data directory.

That's not my point: there actually is a python package for threejs, under the sagemath organization, and that is a mess, so much that sage-the-distro is not using it.

Moreover, in this case the data directory is not the worst problem, since the package has a versioned directory.


> Contrast with a python package where the search path is implicitly determined by the --libdir used to configure python. Neither sage nor the database package needs to know that libdir for them to work together. Sage depends on conway-polynomials, but conway-polynomials is now completely independent of sage. The dependency graph has become simpler.

Sure, that's nice. But then you do cremona elliptic curves database, and place the files in a python package. Now pari/gp also uses this data, and they like perl more than python, so let us also do a perl package so the database can be found, etc. Now I end up with several copies of the database, several external packages, etc.

> I've been trying to get Sage into Gentoo for 15 years and I'd really like to be able to cross these database packages off my TODO list one-at-a-time. Making them independently useful and of high quality makes that easier, practically and politically.

Sure, for *some* of the datasets, having a high quality independent python package sounds very useful. Also a lot of work to get it right. Now, when you say "cross the database packages off your TODO list", what do you exactly mean? For me it was one of the easier parts of packaging sagemath. I decided that I'd place them in `/usr/share/sagemath` and it's just a matter of dumping files in there. Up until this PR I had to setup `SAGE_SHARE`, now they will be autodetected and that's it. It ain't broken, don't fix it. Changing these packages for new packages (python or not) costs a lot of churn for little or no gain. Unless, of course, you come up with a way to share the cremona database between sagemath and pari/gp, and a general way to manage databases, etc. Ideally something that upstream (= source of the data) can manage and maintain on their own, so I actually get a database package from upstream, just like we get ipython from upstream.

**Comment by kiwifb** [comment]
> > I'm looking forward to a time when we can check out sagelib and work on it without all the `build/` stuff.
> 
> I'm about to switch my void package to use `sagemath-standard` from pypi. At this point I think it's doable.
> 
> Now _that_ is a move I would support: splitting `sage-the-distro` from `sagemath-standard`. Have `sage-the-distro` use `sagemath-standard` as upstream, with no special privileges (other than maybe more testing) wrt other distros.
> 

Music to my ears. Note that I have already switch to sagemath-standard, I just have not changed the name yet (but intend to do so at the 10.3 release). And that means all "interactions" between sagemath and sage-the-distro have to be made optional. `sage/misc/package.py` and its content should be a separate package and you should not need it at build time or run time. Its only value should be to sage-the-distro users.

**Comment by mkoeppe** [comment]
> we have sage_setup split off (why?)

Because it is the build system not only of sagemath-standard, but also of the other modularized distributions.

**Comment by mkoeppe** [comment]
> > I'm about to switch my void package to use `sagemath-standard` from pypi. At this point I think it's doable.

+1

> > Now _that_ is a move I would support: splitting `sage-the-distro` from `sagemath-standard`. Have `sage-the-distro` use `sagemath-standard` as upstream, with no special privileges (other than maybe more testing) wrt other distros.

Yes, that's our shared goal, and remaining links such as assumptions about layout within SAGE_LOCAL should be cut. 

By the way, that's also why sage-conf is separate from sagemath-standard. 
(I know that @tornaria does not need to configure anything in his installation of sage, but there are legitimate deployment situations in which sage-conf cannot be a no-op.)

> Music to my ears. Note that I have already switch to sagemath-standard, I just have not changed the name yet (but intend to do so at the 10.3 release). And that means all "interactions" between sagemath and sage-the-distro have to be made optional.

+1

> `sage/misc/package.py` and its content should be a separate package and you should not need it at build time or run time. Its only value should be to sage-the-distro users.

Getting rid of its use at build and run time is #30746, #30848.
When done, I think we can move it to sage-bootstrap (#31662).

**Comment by kiwifb** [comment]
> > we have sage_setup split off (why?)
> 
> Because it is the build system not only of sagemath-standard, but also of the other modularized distributions.

Not sure I want to engage in that conversation as it is off topic to the PR and can only accelerate from there :)
There could be a more solid argument for separation if there were downstream packages using it independently of sage. But in the modularisation effort you need to get a build component early. You could merge it with a "core" component, that is an option.

Versioning, I would like to say it is a mono-repo thing. But eventually we may reach the point where the only difference between some versions of sage_setup will be the VERSION file. At which point decoupling versions may seem attractive.
Of course we have a big example of a mono-repo with packages that have different version numbers: suitesparse. And it is a nightmare to keep everything in sync as separate packages. But there are two major differences with sage:
* suitesparse has not always been mono repo. Several things just coalesced under one maintainer
* suitesparse is distributed as a single giant tarball while modular sagemath provides individual sdists on pypi.
The later makes it more straightforward to mix and match versions if you so wish. If there are no actual differences between sage_setup-x and sage_setup-x+1 may be I do not have to bother rev-bumping it.

**Comment by mkoeppe** [comment]
> in the modularisation effort you need to get a build component early. You could merge it with a "core" component, that is an option.

But sage_setup is only needed at build time, never at runtime.

> Versioning, I would like to say it is a mono-repo thing. But eventually we may reach the point where the only difference between some versions of sage_setup will be the VERSION file. At which point decoupling versions may seem attractive.

Of course. I'm open to that when it reaches this point.

**Comment by kiwifb** [comment]
> > in the modularisation effort you need to get a build component early. You could merge it with a "core" component, that is an option.
> 
> But sage_setup is only needed at build time, never at runtime.
> 

And I must say that point should make it attractive to binary distribution because it becomes a "-dev" package that do not need to be installed for end users.

**Comment by mkoeppe** [comment]
This separation of packages by the "time" that they used, by the way, motivated splitting sage-docbuild from sage-setup.

**Comment by tornaria** [comment]
Could we just
```
ln -sT ../../src/sage_setup pkgs/sagemath-standard/sage_setup
```
with this symlink in place, running setup.py from within `pkgs/sagemath-standard` works just fine without the `PYTHONPATH` hack I was doing before.

Also: would adding the symlink here mean that the actual files will be shipped inside of the `sagemath-standard` sdist that we get from pypi?

Otherwise, I'll just unpack `sage_setup` inside `sagemath-standard` before building, but it would be much nicer to include it. I cannot and won't use `sage_setup` as a system package until its version is decoupled from `sagemath-standard` version.

**Comment by mkoeppe** [comment]
> I cannot and won't use `sage_setup` as a system package until its version is decoupled from `sagemath-standard` version.

I doubt the "cannot".

**Comment by tornaria** [comment]
> > I cannot and won't use `sage_setup` as a system package until its version is decoupled from `sagemath-standard` version.
> 
> I doubt the "cannot".

Since you expect me to change `sage_setup` for every beta release, sure, I cannot have a system package of `sage_setup`, because it would not work to build sagemath-standard. I would have to be switching the version of `sage_setup` in the system, etc.

Also, what is so special about `sage_setup`? From the top of my head I imagine (a) build cython extensions in a way that works for sagemath (b) install the sagemath jupyter kernel. Now if (a) is to avoid shortcomings of cython build extension, why not try to push those upstream instead of diverging? On (b) I'd rather do it differently than it's done now, but in any case the sagemath jupyter kernel will be shipped by a single distribution, so the code to install that belongs only to that distribution setup.py.

**Comment by mkoeppe** [comment]
> Since you expect me to change `sage_setup` for every beta release, sure, I cannot have a system package of `sage_setup`, because it would not work to build sagemath-standard. I would have to be switching the version of `sage_setup` in the system, etc.

Yes, you ship the new version of sage_setup, then you ship the new versions of the distributions that use it as a build system.

**Comment by kiwifb** [comment]
> Otherwise, I'll just unpack `sage_setup` inside `sagemath-standard` before building, but it would be much nicer to include it. I cannot and won't use `sage_setup` as a system package until its version is decoupled from `sagemath-standard` version.

Like Matthias, I would say it is not a matter of "cannot". But what you describe is doable. One of the reasons I fully switched to sdist even on the development branches for beta is because I made an an ebuild for sage_setup. If I just used a git checkout of the development branch I would use the sage_setup from the git checkout instead of my installed one unless I deleted it from the checkout before starting to build. Now that I exclusively use sdists, I have no shadowing issues. If you do not want to use a separate sage_setup, I would recommend you stick with git checkout and just use master for stable releases.

**Comment by kiwifb** [comment]
I will add that I am further down drinking that kool aid than sage developer themselves. I use sdist for everything while sage's meta build just still does the monolithic things from the git checkout.

**Comment by tornaria** [comment]
> > Since you expect me to change `sage_setup` for every beta release, sure, I cannot have a system package of `sage_setup`, because it would not work to build sagemath-standard. I would have to be switching the version of `sage_setup` in the system, etc.
> 
> Yes, you ship the new version of sage_setup, then you ship the new versions of the distributions that use it as a build system.

No, because we don't ship prereleases. I'm not building betas for the system, I'm building betas for myself, to develop and test. I need to be able to switch from one version to another, etc. and not change anything on my system. So either `sage_setup` supports building a range of versions of `sagemath-standard`, or else it is shipped inside `sagemath-standard`.

**Comment by mkoeppe** [comment]
> I need to be able to switch from one version to another, etc. and not change anything on my system. So either `sage_setup` supports building a range of versions of `sagemath-standard`, or else it is shipped inside `sagemath-standard`.

This seems a rather unusual, rigid constraint. Can you point us to some policy document that would explain that?

**Comment by tornaria** [comment]
> This seems a rather unusual, rigid constraint. Can you point us to some policy document that would explain that?

Let's just stop this off-topic here, before the fireworks begin.

It seems this PR is ready to go, can we get the positive review now?

**Review by kiwifb — APPROVED** [review]

File: src/sage/features/threejs.py
Comment (orlitzky) [reply src/sage/features/threejs.py:35]: > Sure, that's nice. But then you do cremona elliptic curves database, and place the files in a python package. Now pari/gp also uses this data, and they like perl more than python, so let us also do a perl package so the database can be found, etc. Now I end up with several copies of the database, several external packages, etc.

In the worse case (no sharing), you get the same number of copies as if you'd bundled everything with sage.

The database packages don't have to be python packages if that's not the best option. In your example, if we ship the raw data with only a pkg-config file and "build" it with autotools, then both Sage and PARI can query `$datadir` using pkg-config. If the interface is more complicated and has multiple consumers, we could write it in C as opposed to python.

> Sure, for _some_ of the datasets, having a high quality independent python package sounds very useful. Also a lot of work to get it right. Now, when you say "cross the database packages off your TODO list", what do you exactly mean? For me it was one of the easier parts of packaging sagemath. I decided that I'd place them in `/usr/share/sagemath` and it's just a matter of dumping files in there. Up until this PR I had to setup `SAGE_SHARE`, now they will be autodetected and that's it. It ain't broken, don't fix it.

As they are now, these packages are useless without Sage, and depend on paths used by Sage. But we don't have Sage in Gentoo, and it could still be a while before we do. Gentoo has hundreds of developers with teams dedicated to various package ecosystems (e.g. math packages), a QA process, two or three permanently-running tinderboxes, tree cleaners, distfile mirrors, binary package builders...

If I were to add these database packages now, it would be adding unusable junk to the tree that would create work for all of those people, potentially for years. Technically I could do it (anyone can commit a new package), but I would rather not spend my political capital that way, especially for something that I don't think is a great idea in the first place.

**Comment by tornaria** [comment]
Rebased without changes on top of 10.3.beta5.

**Pull request merged**
